gensys/germinate/man/germinate.1
.Dd May 27, 2005
.Os Ubuntu
.ds volume-operating-system Ubuntu
.Dt GERMINATE 1
.Sh NAME
.Nm germinate
.Nd expand dependencies in a list of seed packages
.Sh SYNOPSIS
.Nm
.Op Fl S Ar source
.Op Fl s Ar dist
.Op Fl m Ar mirror
.Op Fl d Ar dist Ns \&,...
.Op Fl a Ar arch
.Op Fl c Ar component Ns \&,...
.Op Fl i
.Op Fl Fl bzr
.Op Fl Fl no\-rdepends
.Sh DESCRIPTION
.Nm
is a program to help with the maintenance of large software distributions.
It takes a list of seed packages and a mirror of the distribution, and
produces outputs with the seed packages and their dependencies and
build-dependencies expanded out in full.
.Ss Seeds
The contents of the Ubuntu distribution, and others, are managed by means of
.Em seeds .
At their simplest, these are lists of packages which are considered
important to have in the main component of the distribution, without
explicitly listing all their dependencies and build-dependencies.
.Pp
Seed lists are typically divided up by category: a
.Li base
or
.Li minimal
seed might list the core set of packages required to make the system run at
all, while a
.Li desktop
seed might list the set of packages installed as part of a default desktop
installation.
.Nm
takes these seeds, adds their dependency trees, and produces an
.Em output
for each seed which contains a dependency-expanded list of package names.
These outputs may be handed on to archive maintenance or CD-building tools.
.Pp
Some seeds may
.Em inherit
from other seeds: they rely on those seeds to be installed.
For example, a
.Li desktop
seed will typically inherit from a
.Li minimal
seed.
.Nm
understands these inheritance relationships.
If a package in the
.Li desktop
seed depends on
.Sq foo ,
but
.Sq foo
is already part of the
.Li minimal
seed or dependency list, then
.Sq foo
will not be added to the
.Li desktop
output.
.Pp
Seeds are stored in text files downloaded from a given URL.
Lines not beginning with
.Sq "\ *\ "
(wiki-style list markup) are ignored.
Seed entries beginning with
.Sq %
expand to all binaries from the given source package.
Seed entries may be followed with
.Sq " [" Ns Ar arch1 Ar arch2 ... Ns \&]
to indicate that they should only be used on the given architectures, or with
.Sq " [!" Ns Ar arch1 No ! Ns Ar arch2 ... Ns \&]
to indicate that they should not be used on the given architectures.
.Pp
Seed entries in parentheses indicate that the seed should be treated as a
recommendation of metapackages generated from this seed, rather than as a
dependency.
.Pp
Seed entries beginning with
.Sq \&!
cause the given package to be blacklisted from the given seed and any seeds
from which it inherits; this may be followed by
.Sq %
as above to blacklist all binaries from the given source package.
Note that this may result in uninstallable packages whose dependencies have
been blacklisted, so use this feature sparingly.
.Pp
A
.Pa STRUCTURE
file alongside the seeds lists their inheritance relationships.
.Ss Build-dependencies and Sq supported
There is typically no need for a default desktop installation to contain all
the compilers and development libraries needed to build itself from source;
if nothing else, it would consume much more space.
Nevertheless, it is normally a requirement for the maintainers of a
distribution to support all the packages necessary to build that
distribution.
.Pp
.Nm
therefore does not add all the packages that result from following
build-dependencies of seed packages and of their dependencies (the
.Dq build-dependency tree )
to every output, unless they are also in the seed or in the dependency list.
Instead, it adds them to the output for the last seed in the
.Pa STRUCTURE
file, conventionally called
.Li supported .
.Pp
Like any other seed, the supported seed may contain its own list of
packages.
It is common to provide support for many software packages which are not in
the default installation, such as debugging libraries, optimised kernels,
alternative language support, and the like.
.Ss Outputs
The output files are named after the seed to which they correspond.
An additional output file is needed for supported, namely
.Sq supported+build\-depends ,
which contains the supported list and the build-depends lists of the other
seeds all joined together.
An
.Sq all
output is produced to represent the entire archive.
.Pp
Some other files are produced for occasional use by experts.
See the
.Pa README
file for full details on these.
.Sh OPTIONS
.Bl -tag -width 4n
.It Xo Fl S ,
.Fl Fl seed\-source Ar source Ns \&,...
.Xc
Fetch seeds from the specified sources.
The default is
.Pa http://people.ubuntu.com/~ubuntu-archive/seeds/ ,
or
.Pa http://bazaar.launchpad.net/~ubuntu-core-dev/ubuntu-seeds/
if the
.Fl Fl bzr
option is used.
.It Xo Fl s ,
.Fl Fl seed\-dist Ar dist
.Xc
Fetch seeds for distribution
.Ar dist .
The default is
.Li ubuntu.hardy .
.It Xo Fl m ,
.Fl Fl mirror Ar mirror
.Xc
Get package lists from
.Ar mirror .
The default is
.Pa http://archive.ubuntu.com/ubuntu/ .
.It Xo Fl d ,
.Fl Fl dist Ar dist Ns \&,...
.Xc
Operate on the specified distributions.
The default is
.Li hardy .
Listing multiple distributions may be useful, for example, when examining
both a released distribution and its security updates.
.It Xo Fl a ,
.Fl Fl arch Ar arch
.Xc
Operate on architecture
.Ar arch .
The default is
.Li i386 .
.It Xo Fl c ,
.Fl Fl components Ar component Ns \&,...
.Xc
Operate on the specified components.
The default is
.Li main .
.It Xo Fl i ,
.Fl Fl ipv6
.Xc
Check IPv6 status of source packages, using the statistics available at
.Pa http://debdev.fabbione.net/stat/ .
.It Fl Fl bzr
Check out seeds from the
.Ic bzr
branch found at
.Ar seed\-source Ns / Ns Ar seed\-dist
rather than fetching them directly from a URL.
Requires
.Ic bzr
to be installed.
.It Fl Fl no\-rdepends
Disable reverse-dependency calculations.
These calculations cause a large number of small files to be written out in
the
.Pa rdepends/
directory, and may take some time.
.It Fl Fl seed\-packages Ar parent Ns / Ns Ar pkg Ns \&,...
Treat each
.Ar pkg
as a seed by itself, inheriting from
.Ar parent
(i.e. assuming that all packages in the
.Ar parent
seed are already installed while calculating the additional dependencies of
.Ar pkg ) .
This allows the use of
.Nm
to calculate the dependencies of individual extra packages.
.El
.Sh BUGS
The wiki-style markup in seeds was inherited from an early implementation,
and is a wart.
.Pp
.Nm
can sometimes be confused by complicated situations involving the order in
which it encounters dependencies on virtual packages.
Explicit entries in seeds may be required to work around this.
.Pp
Handling of installer packages (udebs) is complicated, poorly documented,
and doesn't always work quite right: in particular, packages aren't demoted
to the supported seed when they should be.
.Sh AUTHORS
.An Scott James Remnant Aq scott@canonical.com
.An Colin Watson Aq cjwatson@canonical.com
.Pp
.An -nosplit
.Nm
is copyright \(co 2004, 2005, 2006, 2007, 2008
.An Canonical Ltd .
See the GNU General Public License version 2 or later for copying
conditions.
A copy of the GNU General Public License is available in
.Pa /usr/share/common\-licenses/GPL .


gensys/germinate/germinate.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
"""Update list files from the Wiki."""

# Copyright (c) 2004, 2005, 2006, 2007, 2008 Canonical Ltd.
#
# Germinate is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2, or (at your option) any
# later version.
#
# Germinate is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Germinate; see the file COPYING.  If not, write to the Free
# Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
# 02110-1301, USA.

import gzip
import os
import shutil
import sys
import urllib2
import getopt
import logging
import codecs
import cStringIO

try:
    set # introduced in 2.4
except NameError:
    import sets
    set = sets.Set

import apt_pkg

from Germinate import Germinator
import Germinate.Archive
import Germinate.seeds
import Germinate.version


# Where do we get up-to-date seeds from?
# SEEDS = ["http://people.ubuntu.com/~ubuntu-archive/seeds/"]
# SEEDS_BZR = ["http://bazaar.launchpad.net/~ubuntu-core-dev/ubuntu-seeds/"]
SEEDS = ["http://gensys/ubuntu/seeds/"]
SEEDS_BZR = ["http://gensys/ubuntu/seeds/"]
RELEASE = "ubuntu.hardy"

# If we need to download Packages.gz and/or Sources.gz, where do we get
# them from?
# MIRROR = "http://archive.ubuntu.com/ubuntu/"
MIRROR = "http://gensys/ubuntu/"
SOURCE_MIRROR = None
DIST = ["hardy"]
COMPONENTS = ["main", "restricted"]
ARCH = "i386"

CHECK_IPV6 = False

# If we need to download a new IPv6 dump, where do we get it from?
IPV6DB = "http://debdev.fabbione.net/stat/"


def open_ipv6_tag_file(filename):
    """Download the daily IPv6 db dump if needed, and open it."""
    if os.path.exists(filename):
        return open(filename, "r")

    print "Downloading", filename, "file ..."
    url = IPV6DB + filename + ".gz"
    url_f = urllib2.urlopen(url)
    url_data = cStringIO.StringIO(url_f.read())
    url_f.close()
    print "Decompressing", filename, "file ..."
    gzip_f = gzip.GzipFile(fileobj=url_data)
    f = open(filename, "w")
    for line in gzip_f:
        print >>f, line,
    f.close()
    gzip_f.close()
    url_data.close()

    return open(filename, "r")

def write_list(whyname, filename, g, pkgset):
    pkglist = list(pkgset)
    pkglist.sort()

    pkg_len = len("Package")
    src_len = len("Source")
    why_len = len("Why")
    mnt_len = len("Maintainer")

    for pkg in pkglist:
        _pkg_len = len(pkg)
        if _pkg_len > pkg_len: pkg_len = _pkg_len

        _src_len = len(g.packages[pkg]["Source"])
        if _src_len > src_len: src_len = _src_len

        _why_len = len(g.why[whyname][pkg][0])
        if _why_len > why_len: why_len = _why_len

        _mnt_len = len(g.packages[pkg]["Maintainer"])
        if _mnt_len > mnt_len: mnt_len = _mnt_len

    size = 0
    installed_size = 0

    pkglist.sort()
    f = codecs.open(filename, "w", "utf8", "replace")
    print >>f, "%-*s | %-*s | %-*s | %-*s | %-15s | %-15s" % \
          (pkg_len, "Package",
           src_len, "Source",
           why_len, "Why",
           mnt_len, "Maintainer",
           "Deb Size (B)",
           "Inst Size (KB)")
    print >>f, ("-" * pkg_len) + "-+-" + ("-" * src_len) + "-+-" \
          + ("-" * why_len) + "-+-" + ("-" * mnt_len) + "-+-" \
          + ("-" * 15) + "-+-" + ("-" * 15) + "-"
    for pkg in pkglist:
        size += g.packages[pkg]["Size"]
        installed_size += g.packages[pkg]["Installed-Size"]
        print >>f, "%-*s | %-*s | %-*s | %-*s | %15d | %15d" % \
              (pkg_len, pkg,
               src_len, g.packages[pkg]["Source"],
               why_len, g.why[whyname][pkg][0],
               mnt_len, g.packages[pkg]["Maintainer"],
               g.packages[pkg]["Size"],
               g.packages[pkg]["Installed-Size"])
    print >>f, ("-" * (pkg_len + src_len + why_len + mnt_len + 9)) + "-+-" \
          + ("-" * 15) + "-+-" + ("-" * 15) + "-"
    print >>f, "%*s | %15d | %15d" % \
          ((pkg_len + src_len + why_len + mnt_len + 9), "",
           size, installed_size)

    f.close()

def write_source_list(filename, g, srcset):
    global CHECK_IPV6

    srclist = list(srcset)
    srclist.sort()

    src_len = len("Source")
    mnt_len = len("Maintainer")
    ipv6_len = len("IPv6 status")

    for src in srclist:
        _src_len = len(src)
        if _src_len > src_len: src_len = _src_len

        _mnt_len = len(g.sources[src]["Maintainer"])
        if _mnt_len > mnt_len: mnt_len = _mnt_len

        if CHECK_IPV6:
            _ipv6_len = len(g.sources[src]["IPv6"])
            if _ipv6_len > ipv6_len: ipv6_len = _ipv6_len

    srclist.sort()
    f = codecs.open(filename, "w", "utf8", "replace")

    format = "%-*s | %-*s"
    header_args = [src_len, "Source", mnt_len, "Maintainer"]
    separator = ("-" * src_len) + "-+-" + ("-" * mnt_len) + "-"
    if CHECK_IPV6:
        format += " | %-*s"
        header_args.extend((ipv6_len, "IPv6 status"))
        separator += "+-" + ("-" * ipv6_len) + "-"

    print >>f, format % tuple(header_args)
    print >>f, separator
    for src in srclist:
        args = [src_len, src, mnt_len, g.sources[src]["Maintainer"]]
        if CHECK_IPV6:
            args.extend((ipv6_len, g.sources[src]["IPv6"]))
        print >>f, format % tuple(args)

    f.close()

def write_rdepend_list(filename, g, pkg):
    f = open(filename, "w")
    print >>f, pkg
    _write_rdepend_list(f, g, pkg, "", done=set())
    f.close()

def _write_rdepend_list(f, g, pkg, prefix, stack=None, done=None):
    if stack is None:
        stack = []
    else:
        stack = list(stack)
        if pkg in stack:
            print >>f, prefix + "! loop"
            return
    stack.append(pkg)

    if done is None:
        done = set()
    elif pkg in done:
        print >>f, prefix + "! skipped"
        return
    done.add(pkg)

    for seed in g.seeds:
        if pkg in g.seed[seed]:
            print >>f, prefix + "*", seed.title(), "seed"

    if "Reverse-Depends" not in g.packages[pkg]:
        return

    for field in ("Pre-Depends", "Depends",
                  "Build-Depends", "Build-Depends-Indep"):
        if field not in g.packages[pkg]["Reverse-Depends"]:
            continue

        i = 0
        print >>f, prefix + "*", "Reverse", field + ":"
        for dep in g.packages[pkg]["Reverse-Depends"][field]:
            i += 1
            print >>f, prefix + " +- " + dep
            if field.startswith("Build-"):
                continue

            if i == len(g.packages[pkg]["Reverse-Depends"][field]):
                extra = "    "
            else:
                extra = " |  "
            _write_rdepend_list(f, g, dep, prefix + extra, stack, done)

def write_prov_list(filename, provdict):
    provides = provdict.keys()
    provides.sort()

    f = open(filename, "w")
    for prov in provides:
        print >>f, prov

        provlist = list(provdict[prov])
        provlist.sort()
        for pkg in provlist:
            print >>f, "\t%s" % (pkg,)
        print >>f
    f.close()

def write_structure(filename, structure):
    f = open(filename, "w")
    for line in structure:
        print >>f, line
    f.close()

def write_seedtext(filename, seedtext):
    f = open(filename, "w")
    for line in seedtext:
        print >>f, line.rstrip('\n')


def usage(f):
    print >>f, """Usage: germinate.py [options]

Options:

  -h, --help            Print this help message and exit.
  --version             Output version information and exit.
  -v, --verbose         Be more verbose when processing seeds.
  -S, --seed-source=SOURCE
                        Fetch seeds from SOURCE
                        (default: %s).
  -s, --seed-dist=DIST  Fetch seeds for distribution DIST (default: %s).
  -m, --mirror=MIRROR   Get package lists from MIRROR
                        (default: %s).
  --source-mirror=MIRROR
                        Get source package lists from mirror
                        (default: value of --mirror).
  -d, --dist=DIST       Operate on distribution DIST (default: %s).
  -a, --arch=ARCH       Operate on architecture ARCH (default: %s).
  -c, --components=COMPS
                        Operate on components COMPS (default: %s).
  -i, --ipv6            Check IPv6 status of source packages.
  --bzr                 Fetch seeds using bzr. Requires bzr to be installed.
  --cleanup             Don't cache Packages or Sources files.
  --no-rdepends         Disable reverse-dependency calculations.
  --seed-packages=PARENT/PKG,PARENT/PKG,...
                        Treat each PKG as a seed by itself, inheriting from
                        PARENT.
""" % (",".join(SEEDS), RELEASE, MIRROR, ",".join(DIST), ARCH,
       ",".join(COMPONENTS))


def main():
    global SEEDS, SEEDS_BZR, RELEASE, MIRROR, SOURCE_MIRROR
    global DIST, ARCH, COMPONENTS, CHECK_IPV6
    verbose = False
    bzr = False
    cleanup = False
    want_rdepends = True
    seed_packages = ()
    seeds_set = False

    g = Germinator()

    try:
        opts, args = getopt.getopt(sys.argv[1:], "hvS:s:m:d:c:a:i",
                                   ["help",
                                    "version",
                                    "verbose",
                                    "seed-source=",
                                    "seed-dist=",
                                    "mirror=",
                                    "source-mirror=",
                                    "dist=",
                                    "components=",
                                    "arch=",
                                    "ipv6",
                                    "bzr",
                                    "cleanup",
                                    "no-rdepends",
                                    "seed-packages="])
    except getopt.GetoptError:
        usage(sys.stderr)
        sys.exit(2)

    for option, value in opts:
        if option in ("-h", "--help"):
            usage(sys.stdout)
            sys.exit()
        elif option == "--version":
            print "%s %s" % (os.path.basename(sys.argv[0]),
                             Germinate.version.VERSION)
            sys.exit()
        elif option in ("-v", "--verbose"):
            verbose = True
        elif option in ("-S", "--seed-source"):
            SEEDS = value.split(",")
            seeds_set = True
        elif option in ("-s", "--seed-dist"):
            RELEASE = value
        elif option in ("-m", "--mirror"):
            MIRROR = value
            if not MIRROR.endswith("/"):
                MIRROR += "/"
        elif option == "--source-mirror":
            SOURCE_MIRROR = value
            if not SOURCE_MIRROR.endswith("/"):
                SOURCE_MIRROR += "/"
        elif option in ("-d", "--dist"):
            DIST = value.split(",")
        elif option in ("-c", "--components"):
            COMPONENTS = value.split(",")
        elif option in ("-a", "--arch"):
            ARCH = value
        elif option in ("-i", "--ipv6"):
            CHECK_IPV6 = True
        elif option == "--bzr":
            bzr = True
            if not seeds_set:
                SEEDS = SEEDS_BZR
        elif option == "--cleanup":
            cleanup = True
        elif option == "--no-rdepends":
            want_rdepends = False
        elif option == "--seed-packages":
            seed_packages = value.split(',')

    logger = logging.getLogger()
    if verbose:
        logger.setLevel(logging.DEBUG)
    else:
        logger.setLevel(Germinator.PROGRESS)
    handler = logging.StreamHandler(sys.stdout)
    handler.setFormatter(logging.Formatter('%(levelname)s%(message)s'))
    logger.addHandler(handler)

    apt_pkg.InitConfig()
    apt_pkg.Config.Set("APT::Architecture", ARCH)

    Germinate.Archive.TagFile(MIRROR, SOURCE_MIRROR).feed(
        g, DIST, COMPONENTS, ARCH, cleanup)

    if CHECK_IPV6:
        g.parseIPv6(open_ipv6_tag_file("dailydump"))

    if os.path.isfile("hints"):
        g.parseHints(open("hints"))

    blacklist = Germinate.seeds.open_seed(SEEDS, RELEASE, "blacklist", bzr)
    if blacklist is not None:
        g.parseBlacklist(blacklist)

    seednames, seedinherit, seedbranches = g.parseStructure(
        SEEDS, RELEASE, bzr)
    seedtexts = {}
    for seedname in seednames:
        seed_fd = Germinate.seeds.open_seed(SEEDS, seedbranches, seedname, bzr)
        seedtexts[seedname] = seed_fd.readlines()
        seed_fd.close()
        g.plantSeed(seedtexts[seedname],
                    ARCH, seedname, list(seedinherit[seedname]), RELEASE)
    for seed_package in seed_packages:
        (parent, pkg) = seed_package.split('/')
        g.plantSeed([" * " + pkg], ARCH, pkg,
                    seedinherit[parent] + [parent], RELEASE)
        seednames.append(pkg)
    g.prune()
    g.grow()
    g.addExtras(RELEASE)
    if want_rdepends:
        g.reverseDepends()

    seednames_extra = list(seednames)
    seednames_extra.append('extra')
    for seedname in seednames_extra:
        write_list(seedname, seedname,
                   g, set(g.seed[seedname]) | set(g.seedrecommends[seedname]) |
                      set(g.depends[seedname]))
        write_list(seedname, seedname + ".seed",
                   g, g.seed[seedname])
        write_list(seedname, seedname + ".seed-recommends",
                   g, g.seedrecommends[seedname])
        write_list(seedname, seedname + ".depends",
                   g, g.depends[seedname])
        write_list(seedname, seedname + ".build-depends",
                   g, g.build_depends[seedname])

        if seedname != "extra":
            write_seedtext(seedname + ".seedtext", seedtexts[seedname])
            write_source_list(seedname + ".sources",
                              g, g.sourcepkgs[seedname])
        write_source_list(seedname + ".build-sources",
                          g, g.build_sourcepkgs[seedname])

    all = set()
    sup = set()
    all_srcs = set()
    sup_srcs = set()
    for seedname in seednames:
        all.update(g.seed[seedname])
        all.update(g.seedrecommends[seedname])
        all.update(g.depends[seedname])
        all.update(g.build_depends[seedname])
        all_srcs.update(g.sourcepkgs[seedname])
        all_srcs.update(g.build_sourcepkgs[seedname])

        if seedname == g.supported:
            sup.update(g.seed[seedname])
            sup.update(g.seedrecommends[seedname])
            sup.update(g.depends[seedname])
            sup_srcs.update(g.sourcepkgs[seedname])

        # Only include those build-dependencies that aren't already in the
        # dependency outputs for inner seeds of supported. This allows
        # supported+build-depends to be usable as an "everything else"
        # output.
        build_depends = dict.fromkeys(g.build_depends[seedname], True)
        build_sourcepkgs = dict.fromkeys(g.build_sourcepkgs[seedname], True)
        for seed in g.innerSeeds(g.supported):
            build_depends.update(dict.fromkeys(g.seed[seed], False))
            build_depends.update(dict.fromkeys(g.seedrecommends[seed], False))
            build_depends.update(dict.fromkeys(g.depends[seed], False))
            build_sourcepkgs.update(dict.fromkeys(g.sourcepkgs[seed], False))
        sup.update([k for (k, v) in build_depends.iteritems() if v])
        sup_srcs.update([k for (k, v) in build_sourcepkgs.iteritems() if v])

    write_list("all", "all", g, all)
    write_source_list("all.sources", g, all_srcs)

    write_list("all", "%s+build-depends" % g.supported, g, sup)
    write_source_list("%s+build-depends.sources" % g.supported, g, sup_srcs)

    write_list("all", "all+extra", g, g.all)
    write_source_list("all+extra.sources", g, g.all_srcs)

    write_prov_list("provides", g.pkgprovides)

    write_structure("structure", g.structure)

    if os.path.exists("rdepends"):
        shutil.rmtree("rdepends")
    if want_rdepends:
        os.mkdir("rdepends")
        os.mkdir(os.path.join("rdepends", "ALL"))
        for pkg in g.all:
            dirname = os.path.join("rdepends", g.packages[pkg]["Source"])
            if not os.path.exists(dirname):
                os.mkdir(dirname)

            write_rdepend_list(os.path.join(dirname, pkg), g, pkg)
            os.symlink(os.path.join("..", g.packages[pkg]["Source"], pkg),
                       os.path.join("rdepends", "ALL", pkg))

    g.writeBlacklisted("blacklisted")

if __name__ == "__main__":
    main()
